### DISCUSSION TITLE
Best practice for testing Nova

### DISCUSSION THREAD

**Discussion by chrillep** [laravel/nova-issues#3743] [Q&A] [answered]
This is a very general question 😄  .
I couldn't find anything about testing Nova in the [docs](https://nova.laravel.com/docs/) .
What are the the best practices for testing 
Nova
- actions
- fields
- filters
- lenses
- metrics
- resources

with
- e2e
- feature 
- unit

ref: 
- https://github.com/archtechx/tenancy/discussions/918

**Comment by chrillep** [comment]
found 

- https://novapackages.com/packages/joshgaber/novaunit - @joshgaber not sure if this will break in v4
- https://github.com/laravel/nova-dusk-suite
   - uses https://github.com/orchestral/testbench-dusk
- https://www.masteringnova.com/
- https://github.com/dillingham/nova-assertions @dillingham not sure if this will break in v4

**Comment by davidhemphill** [comment, marked as answer]
One of the best ways to learn how to test Nova is by checking out our own test suite. We typically use a mixture of unit tests and feature tests (ours are powered by https://github.com/orchestral/testbench, but you can just use Laravel's built-in feature tests for the same purpose).

**Reply by chrillep** [reply]
I'll keep chalking down links until I find an actual tutorial and or best way to do this. 😊

albeit your link has what Is needed technicly.

However it doesn't give you anything about how to do it 😊. That is what I am looking for and I reason is needed.

**Reply by chrillep** [reply]
the underlying context, add it to the docs 😭. please 😅
